---
title: "RCTD Tutorial"
author: "Dylan Cable and Gabriel Grajeda"
date: "Feb 12th, 2025"
output:  
  rmarkdown::html_vignette:
vignette: >
  %\VignetteIndexEntry{spatial-transcriptomics}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    out.width = "100%"
)
```

## Introduction

Robust Cell Type Decomposition (RCTD) is a statistical method for decomposing
cell type mixtures in spatial transcriptomics data. In this vignette, we will
use a simulated dataset to demonstrate how you can run RCTD on spatial
transcriptomics data and visualize your results.

We will use the term *pixel*, synonymous with *spot* or *bead*, to refer to any
observation that measures gene expression at a fixed location. We do not use
the term *cell* because, in spatial transcriptomics data, single *pixels* may
contain mixtures of multiple cells. The purpose of RCTD is to estimate the
proportions of different cell types in each pixel of a dataset.

## Setup

```{r setup, message = FALSE}
library(ggplot2)
library(SummarizedExperiment)

library(spacexr)
```

## Data Preprocessing

Before running RCTD, we must first load in:

- an annotated reference dataset, which is used to learn cell type profiles.
- a spatial transcriptomics dataset, which contains pixels for deconvolution.

### Reference

The reference is created using the `Reference` constructor function, which takes
the following parameters:

*Required parameters*:

1. `counts`: A matrix (or dgCMatrix) of gene expression counts, with genes as
rows and cells as columns. Row names should be unique gene names, and column
names should be unique cell barcodes. Counts should be untransformed integers.

2. `cell_types`: A factor vector containing cell type annotations for each cell
in the reference (identified by barcode). The factor levels should represent the
possible cell types.

*Optional parameters*:

3. `nUMI`: A numeric vector of total UMI counts per cell (identified by
barcode). If not provided, `nUMI` will be calculated as the column sums of the
counts matrix.

4. `min_UMI`: Minimum UMI count for cells to be included in the reference. Cells
with fewer UMIs will be removed.

5. `n_max_cells`: Maximum number of cells per cell type. If exceeded, cells will
be randomly downsampled to this number.

In practice, your reference could be an annotated dataset from:

- Single-nucleus RNA sequencing (snRNA-seq)
- Single-cell RNA sequencing (scRNA-seq)
- Cell type-specific bulk RNA sequencing

Here, we load counts and cell type annotations from the `rctd_simulation`
object, but in a real-world setting, the reference information may be
distributed across several files or R objects.

```{r reference}
# Load simulated data.
data("rctd_simulation")

# Create Reference object.
reference <- Reference(
    rctd_simulation$reference_counts,
    rctd_simulation$reference_cell_types
)

# Examine reference. (optional)
print(dim(counts(reference))) # Gene expression matrix dimensions
table(cell_types(reference)) # Number of occurrences for each cell type
```
In this vignette, we have a reference dataset of 750 genes for 75 cells. We
find that there are three cell types (`ct1`, `ct2`, and `ct3`), each appearing
25 times in the reference.

### Spatial Transcriptomics Data

Next, we will load the spatial transcriptomics data into a `SpatialRNA` object.
The `SpatialRNA` constructor function takes the following parameters:

*Required parameters*:

1. `coords`: A data frame (or matrix) containing x and y coordinates for each
pixel. Row names must be unique pixel barcodes, and there must be two columns
named `"x"` and `"y"` containing numeric coordinates.

2. `counts`: A matrix (or dgCMatrix) of gene expression counts, with genes as
rows and pixels as columns. Row names should be unique gene names, and column
names should be unique pixel barcodes. Counts should be untransformed integers.

*Optional parameters*:

3. `nUMI`: A numeric vector of total UMI counts per pixel (identified by
barcode). If not provided, nUMI will be calculated as the column sums of the
counts matrix.

4. `use_fake_coords`: Logical, FALSE by default. If true, the `coords` parameter
will be ignored and replaced with a placeholder coords matrix (useful for
testing).

Here, we load coordinates and counts from the `rctd_simulation` object. Again,
in practice, this step will depend on how you store your spatial transcriptomics
data.

```{r spatial, fig.align='center', fig.width=5, fig.height=5, out.width='50%'}
# Create SpatialRNA object.
puck <- SpatialRNA(
    rctd_simulation$spatial_rna_coords,
    rctd_simulation$spatial_rna_counts
)

# Examine puck. (optional)
print(dim(counts(puck))) # Gene expression matrix dimensions
ggplot(coords(puck), aes(x, y)) + geom_point(size = 3) + 
    theme(axis.text = element_blank())
```
<br/>
We find that we have a dataset of 750 genes for 12 pixels, plotted above.

The ground truth for our cell type mixtures is given by
`rctd_simulation$proportions_se`. Note that this ground truth will not be known
in practice---it is what we will need to estimate with RCTD.

```{r trueProportions}
unique(assay(rctd_simulation$proportions_se))
```

We find that there are two distinct cell type mixtures in our dataset:

- 90% `ct1` and 10% `ct2`.
- 20% `ct1` and 40% `ct2` and `ct3` each.

We can visualize these mixtures with the `plot_all_weights` function, which
plots each pixel as a pie chart of cell types. We will return to this function
later, when it's time to visualize our RCTD results.

```{r trueProportionsPlot, fig.height = 5, fig.width = 7}
plot_all_weights(
    rctd_simulation$proportions_se,
    r = 0.05, lwd = 0.5, title = "Ground Truth"
)
```

## Running RCTD

### Creating RCTD Object

The last step before running RCTD is to create an `RCTD` object from our
`SpatialRNA` and `Reference` objects using the `create.RCTD` function. Some
important configuration options include:

- `max_cores`: Number of cores used for parallel processing. We recommend
setting this to at least 4 to improve efficiency. If set to 1, parallel
processing is not used.
- `gene_cutoff`, `fc_cutoff`: Used for filtering genes in platform effect
normalization. `gene_cutoff` filters for average expression and `fc_cutoff`
filters for log-fold-change across cell types.
- `gene_cutoff_reg`, `fc_cutoff_reg`: Similar to above, but for the RCTD step.
- `UMI_min`, `UMI_max`: Minimum and maximum read depth for pixels in the
`SpatialRNA` dataset.
- `MAX_MULTI_TYPES`: Maximum number of cell types per pixel (multi mode only).

```{r createRCTD, message = FALSE}
rctd_config <- create.RCTD(puck, reference, max_cores = 1)
```

### Running RCTD

We are now ready to run RCTD using the `run.RCTD` function. 

RCTD has three modes, each suited for different spatial technologies:

- `"doublet"`: Assigns 1-2 cell types per pixel. Best for high-resolution
technologies like Slide-seq and MERFISH.
- `"full"`: No restrictions on number of cell types. Best for lower-resolution
technologies like 100-micron Visium.
- `"multi"`: Like doublet mode but can assign more cell types (up to
MAX_MULTI_TYPES).

The `doublet_mode` argument is used to specify the RCTD mode. For this vignette,
we'll decompose the data into doublets.

```{r runRCTD, message = FALSE}
results_se <- run.RCTD(rctd_config, doublet_mode = "doublet")
```

## RCTD Results

RCTD returns a `SummarizedExperiment` object containing:

- Two assays:
    1. `weights`: Cell type proportions restricted according to the specified
    mode
    2. `weights_full`: Unrestricted cell type proportions

- Additional rowData including:
    * `x`, `y`: Spatial coordinates for each pixel
    * For doublet mode:
        * `spot_class`: Classification as `singlet`, `doublet_certain`,
        `doublet_uncertain`, or `reject`
        * `first_type`, `second_type`: Predicted cell types
        * Additional metrics like `min_score`, `singlet_score` for advanced users
    * For multi mode:
        * `cell_type_list`: List of cell types per pixel
        * `conf_list`: List of cell types with confident predictions
        * Additional metrics like `min_score` for advanced users

We can view the doublet weights for our data:

```{r doubletWeights}
assay(results_se)
```

And the classifications for our data:

```{r classification}
classification_df <- data.frame(
    spot_class = rowData(results_se)$spot_class,
    first_type = rowData(results_se)$first_type,
    second_type = rowData(results_se)$second_type
)
classification_df
```

We can see that RCTD generally succeeded in identifying the components of the
cell mixtures, although pixels with 90% `ct1` and 10% `ct2` were classified
as singlets. (As an exercise, try changing the `DOUBLET_THRESHOLD` parameter
in `create.RCTD` so that these mixtures will be identified as doublets.)

We can visualize the results of our RCTD run with the `plot_all_weights`
function. Simply pass in the RCTD results (and some aesthetic arguments), and
the cell type proportions will be plotted as pie charts for each pixel.

```{r doubletVisualization, fig.height = 5, fig.width = 7}
plot_all_weights(results_se, r = 0.05, lwd = 0.5, title = "Doublet Results")
```

Note that the mixture of three cell types (on the right) has been restricted
to `ct2` and `ct3`. We can visualize the unrestricted (non-doublet) proportions
by setting `use_full_weights` to `TRUE`:

```{r fullVisualization, fig.height = 5, fig.width = 7}
plot_all_weights(
    results_se, use_full_weights = TRUE,
    r = 0.05, lwd = 0.5, title = "Full Results"
)
```

This is very close to the ground truth!

Finally, we can visualize how the proportion of a specific cell type varies
across space with the `plot_cell_type_weight` function. We show an example with
the dummy cell type `ct1`.

```{r densityVisualization, fig.height = 5, fig.width = 7}
plot_cell_type_weight(
    results_se, "ct1", use_full_weights = TRUE,
    size = 5, title = "Density of ct1"
)
```

You are now ready to apply RCTD to your own spatial transcriptomics datasets!

## Session Information
```{r sessionInfo}
sessionInfo()
```
